### ISSUE TITLE
[FEATURE REQ] History and Bookmark filters

### ISSUE THREAD

**Issue by ObaraEmmanuel** [pesos/browser-history#122]
**Is your feature request related to a problem? Please describe.**
It's normally overwhelming fetching history data from a browser that is actively used. It would help to have functionality that allowed the data to be filtered along criteria such as domain, time, folder (for bookmarks), and even protocol.

**Describe the solution you'd like**
A way to filter and access filtered data programmatically as well as CLI options for specifying filters. A good option would be `-F/--filter` since `-f` is already taken.

**Additional context**
I understand there's already sorting functionality available but even that is not exposed through the CLI.

**Comment by Samyak2** [comment]
This is a good idea, but I feel this will make the API much more complex. For domains we will need to search for the query in every URL, for time we need to support various queries (before, after a specific time; in a given day, month, year; etc.).

Since we already allow the data to be exported in various formats, other tools which are better suited for this job can be used. For example, pandas has a nice API for filtering data and can be used with CSV exported history/bookmarks.

---

One issue that I've been thinking of raising here was that the current API of `browser-history` might be too overwhelming for users who only want the history/bookmarks. For example, every browser has 10+ exposed functions of which only 2 (`fetch_history` and `fetch_bookmarks`) are usually required. (I realised this after watching [Scott Meyers' "The Most Important Design Guideline" talk](https://youtu.be/TdajK_SXwoc))

**Comment by ObaraEmmanuel** [comment]
Yeah, listened to the lecture and saw the same. Most of those functions are implementation details and can be made private. But that is a breaking change.  Also, most of the filter functionality would go to the output object unless you want to filter directly using the SQL query

**Comment by Samyak2** [comment]
> Yeah, listened to the lecture and saw the same. Most of those functions are implementation details and can be made private. But that is a breaking change.

Yes, it shouldn't be a major breaking change since those functions are not supposed to be used directly anyway. Also we do not have a stable 1.0 API yet, so it should be ok.

> Also, most of the filter functionality would go to the output object unless you want to filter directly using the SQL query

Oh right. That seems like a good idea, but we still have to take care of the filter language/format, etc.

**Comment by ObaraEmmanuel** [comment]
> Oh right. That seems like a good idea, but we still have to take care of the filter language/format, etc.

For history we can have filters for:
* domain
* timestamp (year, day, day of week, month filters among others)
* keyword
* protocol

As for the CLI filter format I haven't thought of anything yet. Maybe something like `"year=2020&protocol=http&keyword=python"` or `"year=2020,protocol=http,keyword=python"` passed through the `--filter` option

**Comment by Samyak2** [comment]
> For history we can have filters for:
> 
>     * domain
> 
>     * timestamp (year, day, day of week, month filters among others)
> 
>     * keyword
> 
>     * protocol
> 
> 
> As for the CLI filter format I haven't thought of anything yet. Maybe something like `"year=2020&protocol=http&keyword=python"` or `"year=2020,protocol=http,keyword=python"` passed through the `--filter` option

That latter format ( `"year=2020,protocol=http,keyword=python"`) seems quite good. I thought we'd need to also handle other relational operators (`>`, `<`, `<=` and something like `2015<year<2020`), but this seems sufficient.

Those can be arguments to a `filter` function in `Outputs` which can be called from the CLI.

**Comment by ObaraEmmanuel** [comment]
> I thought we'd need to also handle other relational operators (>, <, <= and something like 2015<year<2020), but this seems sufficient.

I had not thought of this. Relational operators are very necessary for numeric filters. We should include them. Parsing this shouldn't be too hard.

**Comment by Samyak2** [comment]
> I had not thought of this. Relational operators are very necessary for numeric filters. We should include them. Parsing this shouldn't be too hard.

Right, the basic operators shouldn't be too difficult.
